lib/jorb/job.ex

```diff
@@ -1,36 +1,191 @@
 
   @doc false
   defmacro __using__(_opts) do
     quote do
       @behaviour Jorb.Job
 
+      def write_queue(_payload), do: write_queue()
+
+      def write_queue(), do: raise("either write_queue/1 or write_queue/0 must be defined")
+
+      defoverridable Jorb.Job
+
       @doc ~S"""
-      Queue a job to be performed later. Send the name of the enqueueing module along
-      so we know which module to send the params later
+      Queue a job to be performed by this module's `perform/1` function
+      later.
       """
-      @spec perform_async(Poison.Encoder.t()) :: :ok | {:error, any}
-      def perform_async(payload) do
-        # Include who sent the message, so we can figure out who's gotta deal with it later
-        body_payload = %{target: __MODULE__, body: payload}
-        queue_name = queue_name() |> Jorb.Backend.prefix_queue_name()
-        Jorb.backend().enqueue(queue_name, body_payload)
+      @spec enqueue(any, Keyword.t()) :: :ok | {:error, String.t()}
+      def enqueue(payload, opts \\ []), do: Jorb.Job.enqueue(__MODULE__, payload, opts)
+
+      @doc ~S"""
+      Attempt to fetch jobs to do, reading from the first item in
+      `read_queues/0` that has messages.  For each message received,
+      `perform/1` is invoked, deleting the message if the return value
+      is `:ok`.
+      """
+      @spec work(Keyword.t()) :: :ok | {:error, String.t()}
+      def work(opts \\ []), do: Jorb.Job.work(__MODULE__, opts)
+
+      @doc ~S"""
+      Returns a list of one or more child specs for GenServers that
+      execute `work(opts)` forever.
+      """
+      @spec workers(Keyword.t()) :: [:supervisor.child_spec()]
+      def workers(opts \\ []), do: Jorb.Job.workers(__MODULE__, opts)
+    end
+  end
+
+  @doc ~S"""
+  Queue a job to be performed by this module's `perform/1` function
+  later.
+
+  Intended for use through modules that `use Jorb.Job`.
+  """
+  @spec enqueue(atom, any, Keyword.t()) :: :ok | {:error, String.t()}
+  def enqueue(module, payload, opts) do
+    message = %{"target" => module, "body" => payload}
+    queue = module.write_queue(payload)
+    Jorb.config(:backend, [], module).enqueue_message(queue, message, opts)
+  end
+
+  @doc ~S"""
+  Attempt to fetch jobs to do, reading from the first item in
+  `read_queues/0` that has messages.  For each message received,
+  `perform/1` is invoked, deleting the message if the return value
+  is `:ok`.
+
+  Intended for use through modules that `use Jorb.Job`.
+  """
+  @spec workers(atom, Keyword.t()) :: [:supervisor.child_spec()]
+  def workers(module, opts) do
+    1..Jorb.config(:worker_count, opts, module)
+    |> Enum.map(fn i ->
+      %{
+        id: {module, :worker, i},
+        start: {Jorb.Worker, :start_link, [[{:module, module} | opts]]},
+        type: :worker,
+        restart: :permanent,
+        shutdown: 5000
+      }
+    end)
+  end
+
+  @doc ~S"""
+  Returns a list of one or more child specs for GenServers that
```

Comment:
Does this actually? It looks like it just starts reading from the things

Reply:
Unless I don't know what `Task.yield_many` does

Reply:
Oh poop, I switched the docs for `work` and `workers`, good catch.